### PR TITLE
add initialpose to emulation_base

### PR DESCRIPTION
**Pull request by benmaidel**

### CODE REVIEW

File: cob_hardware_emulation/scripts/emulation_base.py [hunk 1]
```diff
@@ -7,17 +7,21 @@
 import tf_conversions
 import tf2_ros
 import actionlib
-from geometry_msgs.msg import Twist, TransformStamped
+from geometry_msgs.msg import Twist, Transform, TransformStamped
+from geometry_msgs.msg import PoseWithCovarianceStamped
 from move_base_msgs.msg import MoveBaseAction, MoveBaseActionResult
 from nav_msgs.msg import Odometry
 
+from tf_conversions import
```
Comment (floweisshardt) [inline cob_hardware_emulation/scripts/emulation_base.py]: syntax error

File: cob_hardware_emulation/scripts/emulation_base.py [hunk 2]
```diff
@@ -26,9 +30,9 @@ def __init__(self):
         # TODO
         # - add service reset odometry
```
Comment (floweisshardt) [inline cob_hardware_emulation/scripts/emulation_base.py]: add service reset odometry (/base/odometry_controller/reset_odometry [std_srvs::Trigger])

File: cob_hardware_emulation/scripts/emulation_base.py [hunk 3]
```diff
@@ -93,12 +110,11 @@ def timer_cb(self, event):
         t_loc.transform.rotation = self.odom.pose.pose.orientation
 
         # pub odom_combined --> map
-        # we emulate 'perfect' odometry, so /odom_combined is always the same as /map
         t_odom = TransformStamped()
         t_odom.header.stamp = rospy.Time.now()
         t_odom.header.frame_id = "map"
         t_odom.child_frame_id = "odom_combined"
-        t_odom.transform.rotation.w = 1.0
+        t_odom.transform. = self.initial_pose
```
Comment (floweisshardt) [inline cob_hardware_emulation/scripts/emulation_base.py]: syntax error

File: cob_hardware_emulation/scripts/emulation_base.py [hunk 4]
```diff
@@ -26,9 +30,9 @@ def __init__(self):
         # TODO
         # - add service reset odometry
         # - speed factor
-        # - add 2D Pose Estimation through topic
 
         rospy.Subscriber("/base/twist_controller/command", Twist, self.twist_callback, queue_size=1)
+        rospy.Subscriber("/initalpose", PoseWithCovarianceStamped, self.initalpose_callback, queue_size=1)
```
Comment (floweisshardt) [inline cob_hardware_emulation/scripts/emulation_base.py]: typo: inital --> init**i**al

**Review by floweisshardt — CHANGES_REQUESTED** [review]
2D pose estimate is not working properly. It currently just sets the transformation from map to odom_combined and does not take base_link to odom_combined into account. The result is that the robot is not localized at the pose which is specified by 2D pose estimate

**Comment by benmaidel** [comment]
> 2D pose estimate is not working properly. It currently just sets the transformation from map to odom_combined and does not take base_link to odom_combined into account. The result is that the robot is not localized at the pose which is specified by 2D pose estimate

simple fix for now, by resetting odometry on initalpose

**Review by floweisshardt — APPROVED** [review]

**Comment by floweisshardt** [comment]
workaround with resetting odometry works for now. Remaining issue is tracked in https://github.com/mojin-robotics/cob4/issues/1281

**Pull request merged**